Suggested conda go package

sodre opened the pull request:
<!--
Thank you for pull request.
Below are a few things we ask you kindly to self-check before getting a review. Remove checks that are not relevant.
-->
Checklist
* [x] Used a fork of the feedstock to propose changes
* [x] Bumped the build number (if the version is unchanged)
* [x] Reset the build number to `0` (if the version changed)
* [x] [Re-rendered]( https://conda-forge.org/docs/conda_smithy.html#how-to-re-render ) with the latest `conda-smithy`
* [x] Ensured the license file is being packaged.

<!--
Please note any issues this fixes using [closing keywords]( https://help.github.com/articles/closing-issues-using-keywords/ ):
-->

<!--
Please add any other relevant info below:
-->

sodre committed:
Use go_compiler

sodre committed:
MNT: Re-rendered with conda-smithy 3.1.6 and pinning 2018.07.01

sodre committed:
Suggested go package structure once compiler is added

sodre committed:
update build number

conda-forge-linter commented:
Hi! This is the friendly automated conda-forge-linting service.

I just wanted to let you know that I linted all conda-recipes in your PR (```recipe```) and found it was in an excellent condition.

sodre commented on .ci_support/linux_.yaml:
```diff
@@ -0,0 +1,2 @@
+go_compiler:
+- go
```
This file would be removed once conda-forge/conda-forge-pinning-feedstock#95 is merged.

sodre replied:
This file is okay here. It was generated by `conda-smithy render.`

sodre commented on recipe/build.sh:
```diff
@@ -1,23 +1,8 @@
-export RUNC_VERSION="0.1.0"
-curl -L https://github.com/opencontainers/runc/archive/v${RUNC_VERSION}.tar.gz | tar -xzC "src/github.com/opencontainers"
-mv "src/github.com/opencontainers/runc-${RUNC_VERSION}" "src/github.com/opencontainers/runc"
-
```
All of that is replaced with the multi-source line in the updated meta.yaml file.

sodre commented on recipe/build.sh:
```diff
@@ -1,23 +1,8 @@
-
 export CGO_ENABLED=0
 export GOARCH="amd64"
 export GOOS="linux"
```
GOARCH and GOOS are set by the `go_{linux,osx,win}-64` packages if you are using conda's `target_platform`

sodre commented on recipe/build.sh:
```diff
@@ -1,23 +1,8 @@
 export GOARCH="amd64"
 export GOOS="linux"
-export GOPATH="$(pwd)"
+export GOPATH=$SRC_DIR
```
If we agree that GOPATH can always be set to $SRC_DIR, then I can update the activate/deactivate scripts in the compiler specific packages.

jakirkham replied:
I think this is a sensible default if you agree. Should we raise an issue over at the `go` feedstock to get feedback from others?

sodre replied:
I raised the issue here https://github.com/conda-forge/go-feedstock/issues/31.

jakirkham commented on recipe/meta.yaml:
```diff
@@ -1,22 +1,27 @@
 {% set version = "1.10" %}
+{% set runc_version = "0.1.0" %}
```
Probably should bump to `0.1.1` while we are at it.

jakirkham replied:
Let's deal with this independently.

jakirkham replied:
Note to self: This comes from [here]( https://github.com/tianon/gosu/blob/d2937f478b317e55a77e0e0ed4f79a333a6f5735/Dockerfile#L5 ). So we probably shouldn't change it unless upstream does.

sodre commented on recipe/build.sh:
```diff
@@ -1,23 +1,8 @@
 go build -v -ldflags '-d -s -w' -o "${PREFIX}/bin/gosu"
-
-mkdir -p "${SRC_DIR}"
-cp "LICENSE" "${SRC_DIR}/LICENSE"
```
License code can be pushed to meta.yaml, thus reducing complexity of build.sh as well.

jakirkham replied:
Yeah, `conda-build` has gotten a lot smarter since this was written. 😉

sodre reviewed (commented):
I annotated my changes to explain why I think this is a good start for coming up with a default example for go packages.

sodre committed:
Remove GOOS and GOARCH

jakirkham commented:
Thanks @sodre. Will give this a closer look later.

Do you have any thoughts on how we should handle the `runc` version here?

sodre commented:
No problem, take your time. I would suggest waiting until conda-forge/conda-forge-pinning-feedstock#95 is merged. 

I can update the runc_version to 0.1.1, but I am not sure what exactly you asking as far as how we should handle it.

jakirkham commented:
Ok, can wait until that is resolved.

Basically `gosu` is vendoring it's dependencies, which doesn't seem quite right. Didn't know if you knew of a better way to solve this in Go. Alternatively we could include `runc`'s version in this package's version. Not sure if that makes sense either though. General thoughts on handling dependencies with Go would be helpful.

Also you might want to update issue ( https://github.com/conda-forge/go-feedstock/issues/2 ) with the recommendations you have added here. It could wait until the `go` compiler stuff is merged if that would be better.

sodre committed:
Update SHA256 location

sodre commented on recipe/build.sh:
```diff
@@ -1,23 +1,6 @@
-
 export CGO_ENABLED=0
-export GOARCH="amd64"
-export GOOS="linux"
```
These are replaced by the activate/deactivate scripts in the `go_linux-64` package.

sodre commented:
Hi @jakirkham, 

I looked at the build code for `gosu`. It is a little awkward that we have to bring `runc` separately, but venturing dependencies is the norm in `go` programs. I believe `gosu` predates tools like `govendor` which make this process simple to maintain. Nowadays, `go` packages such as `kubernetes`, `terraform`, and `etcd` include their vendored dependencies within the source code itself in a `vendor` folder. 

I would not change the version string, but we should make sure that if gosu is ever upgraded to 0.11 that our meta.yaml file matches the `runc` version from their [Dockerfile.](https://github.com/tianon/gosu/blob/master/Dockerfile#L5)

Lastly, there is no point in depending on a future `conda-forge/runc` package since all the go programs I've seen so far are statically compiled.

jakirkham commented:
Was afraid of that. Well I guess we will embrace the vendoring then.

That's a good suggestion. Will keep that in mind. Thanks.

One advantage might be build times, but AFAIK we are not in any danger of going over with Go packages yet. Though feel free to correct me if I'm wrong.

Does this need to be refreshed now that `go` is a compiler or are we still waiting on some things?

sodre commented:
I will refresh this recipe and remove the WIP once I'm done.

jakirkham commented:
Friendly nudge 😉

sodre commented:
@conda-forge-admin, please rerender

conda-forge-admin committed:
MNT: Re-rendered with conda-smithy 3.1.12 and pinning 2018.09.27

sodre commented:
@jakirkham, this is ready to merge. Sorry for the delay.

jakirkham commented on recipe/build.sh:
```diff
@@ -1,23 +1,6 @@
-curl -L https://github.com/opencontainers/runc/archive/v${RUNC_VERSION}.tar.gz | tar -xzC "src/github.com/opencontainers"
-mv "src/github.com/opencontainers/runc-${RUNC_VERSION}" "src/github.com/opencontainers/runc"
-
 export CGO_ENABLED=0
```
Is this still needed? Meaning I know it is needed for this recipe, but not if it is still needed with the `go` compiler.

sodre replied:
I think this is not needed here. I am removing to see if the code still builds.

Here is the best description I've found about the CGO_ENABLED flag.
https://go-review.googlesource.com/c/go/+/12603/2/src/cmd/cgo/doc.go

jakirkham replied:
That review makes it sound like we still need to set `CGO_ENABLED=0` if we expect there to be no cgo usage.

FWIW this comes from [here]( https://github.com/tianon/gosu/blob/d2937f478b317e55a77e0e0ed4f79a333a6f5735/Dockerfile#L14-L15 ). I think they are mainly interested in building portable binaries. Though am not clear how they are not still in some way attached to the C runtime library. So this may be less necessary for us.

That said, I think we can table this for now. Was just curious since you know more about Go than I whether this made sense for us. We can save it for another discussion though.

jakirkham commented:
Thanks @sodre. LGTM.

Had [one comment]( https://github.com/conda-forge/gosu-feedstock/pull/3#discussion_r222023869 ) that we can handle asynchronously and [one non-blocking question above]( https://github.com/conda-forge/gosu-feedstock/pull/3#pullrequestreview-160812820 ). Otherwise I think we should merge.

sodre committed:
!fixup

jakirkham committed:
Drop `GOPATH` as the `go` compiler sets this

jakirkham committed:
Add a newline

Slight style adjustment for the build script.

jakirkham committed:
Template out checksums

As a convenience for maintenance, template the checksums as well.

jakirkham committed:
Add `-d` flag to `go`

This is coming from the Docker build that `gosu` is using. So go ahead
and make use of it ourselves as well.

jakirkham committed:
Use `cd`

As we are not moving around directories a lot, there is no real need to
track history.

jakirkham committed:
Place `folder` first in `source`

Mirrors the old format in `source` with `fn` nicely. Also makes it easy
to see where things are being downloaded, which is our primary concern
in terms of build script maintenance.

jakirkham commented:
@conda-forge-admin, please re-render.

jakirkham committed:
Reorder Jinja variables

Group by package out of convenience.

conda-forge-linter commented:
Hi! This is the friendly automated conda-forge-webservice.

I tried to re-render for you, but it looks like there was nothing to do.

conda-forge-linter commented:
Hi! This is the friendly automated conda-forge-linting service.

I just wanted to let you know that I linted all conda-recipes in your PR (```recipe```) and found it was in an excellent condition.

jakirkham reviewed (approved):
LGTM. Thanks so much for working on this @sodre. Will merge once CI passes.
